faw/ci/ci_container_service.py

```diff
@@ -1307,6 +1578,101 @@ async def _get_decisions(self, req):
     return t
 
 
+def start_devmount_watcher(*, config_dir, build_dir, devmounts, faw_docker_id, builder_thread):
+
+    import watchfiles
+
+    class DevMountWatcher(threading.Thread):
+        def __init__(self):
+            super().__init__(daemon=True)
+            self._stop_event = threading.Event()
+
+            # Compute the set of files we need to look for and
+            # map it to the stages we would need to rebuild in
+            # case they change. We also collect a map from files
+            # to devmounts.
+            self._path_stages_map = collections.defaultdict(set)
+            self._path_devmounts_map = collections.defaultdict(set)
+            for name, info in devmounts.items():
+                if not info.valid:
+                    continue
+
+                all_paths = []
+                for p in info.trigger_paths:
+                    full_path = os.path.join(info.mounted_path, p)
+                    paths = glob.glob(full_path, recursive=True)
+                    all_paths.extend(paths)
+
+                for path in all_paths:
+                    for stage_name in info.dependent_stages:
+                        self._path_stages_map[path].add(stage_name)
+
+                    self._path_devmounts_map[path].add(name)
+
+            logging.info(f"Watching {list(self._path_stages_map.keys())}")
+
+        def stop(self):
+            self._stop_event.set()
+
+        def run(self):
+            for changes in watchfiles.watch(*self._path_stages_map.keys(), yield_on_timeout=True):
```

Comment:
Should there be a per-devmount "cooldown" or delay time for these updates? I'd potentially worry that typing in a watched file could trigger updates that break the build before the change is complete.